### DISCUSSION TITLE
How does tokenization work for the STT models like nova-2 in the API?

### DISCUSSION THREAD

**Discussion by TasseDeCafe** [deepgram/community#817] [General help] [answered]
Hi!

When passing an audio file to the API endpoint of nova-2 speech-to-text model, the returned JSON includes a "words" array. I have noticed that for many languages the words seem to be split by white spaces, but for other languages the tokenization algorithm is a bit mysterious.

For example:
- For Mandarin Chinese: there doesn't seem be any word tokenization, the text is split by character
- For Thai: the API tokenizes the text into semantic words and not character per character.
- For Japanese: unlike Chinese, the API sometimes return multiple characters as a "word" (but the splits are mostly incorrect)
- For Vietnamese: the words are split by syllable.

![CleanShot 2024-06-18 at 11 14 22@2x](https://github.com/deepgram/community/assets/44320355/578e55e9-67b5-431a-ae80-b50f2a764fd5)

(request id of this screenshot: d7556df5-70b0-48ca-bb9d-5f14bfcd401e )

Is the tokenization documented somewhere? I am building a language-learning app, and it would help a lot if I knew how words are tokenized in your backend.

Thank you!

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by GA1** [comment]
Hey, any update on this? @team-deepgram

**Comment by jpvajda** [comment]
@GA1  You can check out this demo https://tokenizer.deepgram.com/ which shows how our Summarization token based model behaves. it can give you  a sense of how tokens are counted. Just keep in mind this is a "best guess estimate" and only for Deepgram Summarization. For other features that use tokens we don't have those models included in this demo yet.

Hopefully this at least can give a sense of how tokens are counted by Deeprgram.

**Reply by GA1** [reply]
Hey @jpvajda, this does not answer the question posed by @TasseDeCafe. Could you have a look at it again? We're interested in what library/algorithm you use for splitting sentences into words.

**Reply by jpvajda** [reply]
@GA1  I don't have that information to share. Sorry we can't answer that for you.

**Reply by TasseDeCafe** [reply]
Hi @jpvajda, thank you for the answer. Unfortunately I'm trying to understand how tokenization works for the data returned by the STT API (nova-2). The summarization tokenizer seems to be similar to the one used by LLMs, but this is different.

**Reply by GA1** [reply]
@jpvajda could you let us know why you can't answer it? We are trying to map our own data with the data returned from deepgram nova-2 api. We are considering switching to another provider because we can't make it work 100%.

**Reply by jpvajda** [reply]
@GA1 Ok,  Let me see if this information that can be shared.

**Reply by jpvajda** [reply]
@GA1  @TasseDeCafe  The best I can provide on this question: 

We build and train our own intelligence models and they have their own proprietary tokenizers. Different models can and do have different tokenizers.

the tokenizer demo I shared does represent an example of how that works.

As for how we identify word boundaries in speech-to-text That's proprietary and I can't really comment on how that's implemented.

**Reply by GA1** [reply]
@jpvajda Thank you for your response. Would it be possible to use it via api? Basically, for a given sentence I want to get word bounderies.

**Reply by jpvajda** [reply]
@GA1  The answer is sort of... 😄  

You do get back the word level breakdown in the `words` array  which might allow you to better understand the boundaries. 

```json
{
    "words": [
        {
            "word": "but",
            "start": 6.0,
            "end": 6.16,
            "confidence": 0.9980469,
            "punctuated_word": "But"
        },
        {
            "word": "what",
            "start": 6.16,
            "end": 6.3999996,
            "confidence": 0.99902344,
            "punctuated_word": "what"
        },
        {
            "word": "about",
            "start": 6.3999996,
            "end": 6.72,
            "confidence": 1.0,
            "punctuated_word": "about"
        }
    ]
}
 ```

**Reply by GA1** [reply]
Thanks @jpvajda, we actually already use this. But our current problem is to receive a similar breakdown when the input is a string (and not audio file/stream).

**Reply by GA1** [reply]
@jpvajda so could you provide an api for dividing a given text into words, just like nova-2 model does?

**Reply by SandraRodgers** [reply, marked as answer]
@GA1 We don't have that feature in our API but you could try sending the text string through our TTS API, turning it into audio, and then try sending it through the STT API, to see the word boundaries. This could be a temporary fix until we finish work on an API like what you are asking for.
